controllers/azuremachine_controller.go

```diff
@@ -109,215 +111,297 @@ func (r *AzureMachineReconciler) SetupWithManager(mgr ctrl.Manager, options cont
+        Namespace: azureMachine.Namespace,
+        Name:      cluster.Spec.InfrastructureRef.Name,
+    }
+    azureCluster := &infrav1.AzureCluster{}
+    if err := r.Client.Get(ctx, azureClusterName, azureCluster); err != nil {
+        r.Recorder.Eventf(azureMachine, corev1.EventTypeNormal, "AzureCluster unavailable", "AzureCluster is not available yet")
+        logger.Info("AzureCluster is not available yet")
+        return reconcile.Result{}, nil
+    }
+
+    logger = logger.WithValues("AzureCluster", azureCluster.Name)
+
+    // Create the cluster scope
+    clusterScope, err := scope.NewClusterScope(scope.ClusterScopeParams{
+        Client:       r.Client,
+        Logger:       logger,
+        Cluster:      cluster,
+        AzureCluster: azureCluster,
+    })
+    if err != nil {
+        r.Recorder.Eventf(azureCluster, corev1.EventTypeWarning, "Error creating the cluster scope", err.Error())
+        return reconcile.Result{}, err
+    }
+
+    //Fetch the AzureIPPool
+    azureIPPoolName := client.ObjectKey{
+        Namespace: "default",
+        Name:      "ase-ip-pool", //get the ip-pool-name from the cluster spec
+    }
+    azureIPPool := &infrav1.AzureIPPool{}
+    if err := r.Client.Get(ctx, azureIPPoolName, azureIPPool); err != nil {
+    // is this event fine?
+    r.Recorder.Eventf(azureMachine, corev1.EventTypeWarning, "Error getting the AzureIPpool", err.Error())
+    return reconcile.Result{}, errors.Wrapf(err, "Failed to retrieve Azure IP pool")
+    }
+
+    logger = logger.WithValues("AzureIPPool", azureIPPool.Name)
+    
+    //Create AzureIPPoolScope
+    azureIPPoolScope, err := scope.NewIPPoolScope(scope.IPPoolScopeParams{
+        Client:       r.Client,
+        Logger:       logger,
+        AzureIPPool: azureIPPool,
+    })
+    if err != nil {
+        r.Recorder.Eventf(azureIPPool, corev1.EventTypeWarning, "Error creating the ip pool scope", err.Error())
+        return reconcile.Result{}, err
+    }
+
+    // Create the machine scope
+    machineScope, err := scope.NewMachineScope(scope.MachineScopeParams{
+        Logger:           logger,
+        Client:           r.Client,
+        Machine:          machine,
+        AzureMachine:     azureMachine,
+        ClusterDescriber: clusterScope,
+    })
+    if err != nil {
+        r.Recorder.Eventf(azureMachine, corev1.EventTypeWarning, "Error creating the machine scope", err.Error())
+        return reconcile.Result{}, errors.Errorf("failed to create scope: %+v", err)
+    }
+
+    // Always close the scope when exiting this function so we can persist any AzureMachine changes.
+    defer func() {
+        conditions.SetSummary(machineScope.AzureMachine,
+            conditions.WithConditions(
+                infrav1.VMRunningCondition,
+            ),
+            conditions.WithStepCounterIfOnly(
+                infrav1.VMRunningCondition,
+            ),
+        )
+
+        if err := machineScope.Close(ctx); err != nil && reterr == nil {
```

Comment:
do you also need to close the azure ip pool scope?

Reply:
Yes, fixing this and testing